File: crates/c-api/examples/remux.c
// MP4 ファイルをリマルチプレックスするサンプルプログラム
//
// 入力 MP4 ファイルを読み込んでデマルチプレックスし、
// すべてのサンプルを新しい MP4 ファイルに書き直すプログラムである
//
// # ビルド
//
// ```bash
// # mp4-rust のプロジェクトルートで libmp4.a をビルド
// cargo build
//
// # remux.c をコンパイル
// cc -o target/debug/remux \
//    -I crates/c-api/include/ \
//    crates/c-api/examples/remux.c \
//    target/debug/libmp4.a
// ```
//
// # 実行方法
// ```bash
// ./target/debug/remux /path/to/input.mp4 /path/to/output.mp4
// ```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "mp4.h"

#define BUFFER_SIZE (1024 * 1024)  // 1MB のバッファサイズ

// サンプルエントリー種別を文字列に変換
const char *get_sample_entry_kind_name(enum Mp4SampleEntryKind kind) {
    switch (kind) {
        case MP4_SAMPLE_ENTRY_KIND_AVC1:
            return "AVC1 (H.264)";
        case MP4_SAMPLE_ENTRY_KIND_HEV1:
            return "HEV1 (H.265/HEVC)";
        case MP4_SAMPLE_ENTRY_KIND_HVC1:
            return "HVC1 (H.265/HEVC)";
        case MP4_SAMPLE_ENTRY_KIND_VP08:
            return "VP08 (VP8)";
        case MP4_SAMPLE_ENTRY_KIND_VP09:
            return "VP09 (VP9)";
        case MP4_SAMPLE_ENTRY_KIND_AV01:
            return "AV01 (AV1)";
        case MP4_SAMPLE_ENTRY_KIND_OPUS:
            return "Opus";
        case MP4_SAMPLE_ENTRY_KIND_MP4A:
            return "MP4A (AAC)";
        default:
            return "Unknown";
    }
}

int main(int argc, char *argv[]) {
    if (argc < 3) {
        fprintf(stderr, "Usage: %s <input_mp4> <output_mp4>\n", argv[0]);
        return 1;
    }

    const char *input_filepath = argv[1];
    const char *output_filepath = argv[2];

    // ==================== デマルチプレックサーのセットアップ ====================
    FILE *input_file = fopen(input_filepath, "rb");
    if (!input_file) {
        fprintf(stderr, "Error: Could not open input file '%s'\n", input_filepath);
        return 1;
    }

    // デマルチプレックサーを作成
    struct Mp4FileDemuxer *demuxer = mp4_file_demuxer_new();
    if (!demuxer) {
        fprintf(stderr, "Error: Could not create demuxer\n");
        fclose(input_file);
        return 1;
    }

    // 読み込み用バッファを割り当て
    uint8_t *read_buffer = (uint8_t *)malloc(BUFFER_SIZE);
    if (!read_buffer) {
        fprintf(stderr, "Error: Could not allocate read buffer\n");
        mp4_file_demuxer_free(demuxer);
        fclose(input_file);
        return 1;
    }

    // ==================== マルチプレックサーのセットアップ ====================
    FILE *output_file = fopen(output_filepath, "wb");
    if (!output_file) {
        fprintf(stderr, "Error: Could not open output file '%s'\n", output_filepath);
        free(read_buffer);
        mp4_file_demuxer_free(demuxer);
        fclose(input_file);
        return 1;
    }

    // マルチプレックサーを作成
    struct Mp4FileMuxer *muxer = mp4_file_muxer_new();
    if (!muxer) {
        fprintf(stderr, "Error: Could not create muxer\n");
        fclose(output_file);
        free(read_buffer);
        mp4_file_demuxer_free(demuxer);
        fclose(input_file);
        return 1;
    }

    // マルチプレックサーを初期化
    enum Mp4Error err = mp4_file_muxer_initialize(muxer);
    if (err != MP4_ERROR_OK) {
        fprintf(stderr, "Error: Failed to initialize muxer: %d\n", err);
        const char *error_msg = mp4_file_muxer_get_last_error(muxer);
        if (error_msg) {
            fprintf(stderr, "  %s\n", error_msg);
        }
        mp4_file_muxer_free(muxer);
        fclose(output_file);
        free(read_buffer);
        mp4_file_demuxer_free(demuxer);
        fclose(input_file);
        return 1;
    }

    // マルチプレックサーの初期出力データを書き込む
    uint64_t output_offset;
    uint32_t output_size;
    const uint8_t *output_data;
    uint64_t current_output_data_offset = 0;

    printf("Writing initial muxer boxes...\n");
    while (mp4_file_muxer_next_output(muxer, &output_offset, &output_size, &output_data) ==
           MP4_ERROR_OK) {
        if (output_size == 0) break;

        if (fseek(output_file, output_offset, SEEK_SET) != 0) {
            fprintf(stderr, "Error: Failed to seek in output file\n");
            goto cleanup;
        }

        if (fwrite(output_data, 1, output_size, output_file) != output_size) {
            fprintf(stderr, "Error: Failed to write to output file\n");
            goto cleanup;
        }

        printf("  Wrote %u bytes at offset %lu\n", output_size, output_offset);

        // 次のサンプルデータの開始位置を追跡
        current_output_data_offset = output_offset + output_size;
    }

    printf("Sample data will start at offset: %lu\n\n", current_output_data_offset);

    // ==================== 入力ファイルのデマルチプレックス ====================
    printf("Demuxing input file...\n");

    // デマルチプレックサーが初期化完了するまで、必要なデータを読み込んで入力する
    while (true) {
        uint64_t required_position;
        int32_t required_size;

        err = mp4_file_demuxer_get_required_input(demuxer, &required_position, &required_size);
        if (err != MP4_ERROR_OK) {
            fprintf(stderr, "Error: Failed to get required input: %d\n", err);
            goto cleanup;
        }

        // 必要なデータサイズが 0 の場合は初期化完了
        if (required_size == 0) {
            break;
        }

        if (fseek(input_file, required_position, SEEK_SET) != 0) {
            fprintf(stderr, "Error: Could not seek to position %lu\n", required_position);
            goto cleanup;
        }

        // 読み込むサイズを決定
        size_t read_size = BUFFER_SIZE;
        if (required_size > 0) {
            // 特定のサイズが要求されている場合
            read_size = (size_t)required_size;
        } else if (required_size == -1) {
            // ファイル末尾までの読み込みが必要な場合
            long current_pos = ftell(input_file);
            fseek(input_file, 0, SEEK_END);
            long file_size = ftell(input_file);
            fseek(input_file, required_position, SEEK_SET);

            read_size = file_size - required_position;
        }

        // バッファサイズを超えていたらエラーにする
        // （実際には、許容可能な範囲内ならバッファをリサイズすべき）
        if (read_size > BUFFER_SIZE) {
            fprintf(stderr, "Error: read_size (%zu) exceeds BUFFER_SIZE (%zu). \n",
                    read_size, (size_t)BUFFER_SIZE);
            goto cleanup;
        }

        size_t bytes_read = fread(read_buffer, 1, read_size, input_file);
        if (bytes_read == 0 && ferror(input_file)) {
            fprintf(stderr, "Error: Failed to read input file\n");
            goto cleanup;
        }

        // デマルチプレックサーにデータを入力
        mp4_file_demuxer_handle_input(demuxer, required_position, read_buffer,
                                      (uint32_t)bytes_read);
    }

    // トラック情報を取得
    const struct Mp4DemuxTrackInfo *tracks;
    uint32_t track_count;

    err = mp4_file_demuxer_get_tracks(demuxer, &tracks, &track_count);
    if (err != MP4_ERROR_OK) {
        fprintf(stderr, "Error: Failed to get tracks: %d\n", err);
        goto cleanup;
    }

    printf("Found %u track(s)\n\n", track_count);

    // ==================== サンプルのリマルチプレックス ====================
    printf("Remuxing samples...\n");

    uint32_t sample_count = 0;
    struct Mp4DemuxSample demux_sample;

    // 時系列順にサンプルを処理
    while (true) {
        err = mp4_file_demuxer_next_sample(demuxer, &demux_sample);

        if (err == MP4_ERROR_NO_MORE_SAMPLES) {
            break;
        }

        if (err != MP4_ERROR_OK) {
            fprintf(stderr, "Error: Failed to get next sample: %d\n", err);
            const char *error_msg = mp4_file_demuxer_get_last_error(demuxer);
            if (error_msg) {
                fprintf(stderr, "  %s\n", error_msg);
            }
            goto cleanup;
        }

        // 入力ファイルからサンプルデータを読み込む
        if (fseek(input_file, demux_sample.data_offset, SEEK_SET) != 0) {
            fprintf(stderr, "Error: Could not seek to sample data offset %lu\n",
                    demux_sample.data_offset);
            goto cleanup;
        }

        // サンプルデータ用のメモリを割り当て
        uint8_t *sample_data = (uint8_t *)malloc(demux_sample.data_size);
        if (!sample_data) {
            fprintf(stderr, "Error: Could not allocate memory for sample data\n");
            goto cleanup;
        }

        // サンプルデータを読み込む
        size_t bytes_read = fread(sample_data, 1, demux_sample.data_size, input_file);
        if (bytes_read != demux_sample.data_size) {
            fprintf(stderr, "Error: Failed to read sample data (expected %zu bytes, got %zu)\n",
                    demux_sample.data_size, bytes_read);
            free(sample_data);
            goto cleanup;
        }

        // サンプルデータを出力ファイルに追記
        if (fwrite(sample_data, 1, demux_sample.data_size, output_file) !=
            demux_sample.data_size) {
            fprintf(stderr, "Error: Failed to write sample data to output file\n");
            free(sample_data);
            goto cleanup;
        }

        free(sample_data);

        // デマルチプレックスサンプルからマルチプレックスサンプルを構築
        struct Mp4MuxSample mux_sample = {
            .track_kind = demux_sample.track->kind,
            .sample_entry = demux_sample.sample_entry,
            .keyframe = demux_sample.keyframe,
            .timescale = demux_sample.track->timescale,
            .duration = demux_sample.duration,
            .data_offset = current_output_data_offset,
            .data_size = (uint32_t)demux_sample.data_size,
        };

        // マルチプレックサーにサンプルを追加
        err = mp4_file_muxer_append_sample(muxer, &mux_sample);
        if (err != MP4_ERROR_OK) {
            fprintf(stderr, "Error: Failed to append sample: %d\n", err);
            const char *error_msg = mp4_file_muxer_get_last_error(muxer);
            if (error_msg) {
                fprintf(stderr, "  %s\n", error_msg);
            }
            goto cleanup;
        }

        sample_count++;

        // 次のサンプルデータの開始位置を更新
        current_output_data_offset += demux_sample.data_size;

        if (sample_count % 100 == 0) {
            printf("  Processed %u samples\n", sample_count);
        }
    }

    printf("Total samples processed: %u\n\n", sample_count);

    // ==================== マルチプレックサーの終了処理 ====================
    printf("Finalizing muxer...\n");

    err = mp4_file_muxer_finalize(muxer);
    if (err != MP4_ERROR_OK) {
        fprintf(stderr, "Error: Failed to finalize muxer: %d\n", err);
        const char *error_msg = mp4_file_muxer_get_last_error(muxer);
        if (error_msg) {
            fprintf(stderr, "  %s\n", error_msg);
        }
        goto cleanup;
    }

    // マルチプレックサーの最終出力データを書き込む
    while (mp4_file_muxer_next_output(muxer, &output_offset, &output_size, &output_data) ==
           MP4_ERROR_OK) {
        if (output_size == 0) break;

        if (fseek(output_file, output_offset, SEEK_SET) != 0) {
            fprintf(stderr, "Error: Failed to seek in output file\n");
            goto cleanup;
        }

        if (fwrite(output_data, 1, output_size, output_file) != output_size) {
            fprintf(stderr, "Error: Failed to write to output file\n");
            goto cleanup;
        }

        printf("  Wrote final %u bytes at offset %lu\n", output_size, output_offset);
    }

    printf("\nSuccessfully remuxed '%s' to '%s'\n", input_filepath, output_filepath);

cleanup:
    // リソースを解放
    mp4_file_muxer_free(muxer);
    fclose(output_file);
    free(read_buffer);
    mp4_file_demuxer_free(demuxer);
    fclose(input_file);

    return 0;
}

File: crates/wasm/examples/remux.js
/**
 * Node.js で mp4-rust の WASM を使って MP4 ファイルをリマルチプレックスする例
 *
 * # 使用方法
 *
 * ```bash
 * cargo build --release --target wasm32-unknown-unknown -p wasm
 * node --experimental-wasm-modules crates/wasm/examples/remux.js /path/to/input.mp4 /path/to/output.mp4
 * ```
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// wasm ファイルを読み込んで初期化する
const wasmPath = path.join(__dirname, '../../../target/wasm32-unknown-unknown/release/mp4_wasm.wasm');
const wasmBuffer = fs.readFileSync(wasmPath);
const wasmInstance = await WebAssembly.instantiate(wasmBuffer);

const {
    memory,
    mp4_file_demuxer_new,
    mp4_file_demuxer_free,
    mp4_file_demuxer_get_required_input,
    mp4_file_demuxer_handle_input,
    mp4_file_demuxer_get_tracks,
    mp4_file_demuxer_next_sample,
    mp4_file_demuxer_get_last_error,
    mp4_file_muxer_new,
    mp4_file_muxer_free,
    mp4_file_muxer_initialize,
    mp4_file_muxer_next_output,
    mp4_file_muxer_append_sample,
    mp4_file_muxer_finalize,
    mp4_file_muxer_get_last_error: mp4_file_muxer_get_last_error_func,
    mp4_mux_sample_from_json,
    mp4_mux_sample_free,
    mp4_alloc,
    mp4_free,
    mp4_vec_ptr,
    mp4_vec_len,
    mp4_vec_free,
} = wasmInstance.instance.exports;

// メモリバッファへのアクセスを簡略化する関数群
function readCString(ptr) {
    const view = new Uint8Array(memory.buffer, ptr);
    let length = 0;
    while (view[length] !== 0) length++;
    return new TextDecoder().decode(view.slice(0, length));
}

function readU8Array(ptr, len) {
    return new Uint8Array(memory.buffer, ptr, len);
}

function readJSON(vecPtr) {
    const ptr = mp4_vec_ptr(vecPtr);
    const len = mp4_vec_len(vecPtr);
    const bytes = readU8Array(ptr, len);
    const jsonStr = new TextDecoder().decode(bytes);
    mp4_vec_free(vecPtr);
    return JSON.parse(jsonStr);
}

function writeToMemory(data) {
    const ptr = mp4_alloc(data.length);
    const view = new Uint8Array(memory.buffer, ptr, data.length);
    view.set(data);
    return ptr;
}

async function remuxMP4File(inputPath, outputPath) {
    if (!fs.existsSync(inputPath)) {
        console.error(`Error: File not found: ${inputPath}`);
        process.exit(1);
    }

    const inputSize = fs.statSync(inputPath).size;
    const inputFile = fs.openSync(inputPath, 'r');
    const outputFile = fs.openSync(outputPath, 'w');

    // デマルチプレックサーを作成
    const demuxerPtr = mp4_file_demuxer_new();
    if (!demuxerPtr) {
        console.error('Error: Could not create demuxer');
        process.exit(1);
    }

    // マルチプレックサーを作成
    const muxerPtr = mp4_file_muxer_new();
    if (!muxerPtr) {
        console.error('Error: Could not create muxer');
        mp4_file_demuxer_free(demuxerPtr);
        fs.closeSync(inputFile);
        fs.closeSync(outputFile);
        process.exit(1);
    }

    console.log(`Remuxing ${path.basename(inputPath)}...`);

    // ==================== デマルチプレックス初期化 ====================

    // ファイルデータを読み込んでデマルチプレックス処理を進める
    while (true) {
        const requiredPosPtr = mp4_alloc(8);
        const requiredSizePtr = mp4_alloc(4);

        const err = mp4_file_demuxer_get_required_input(
            demuxerPtr,
            requiredPosPtr,
            requiredSizePtr
        );

        if (err !== 0) {
            const errorMsg = readCString(mp4_file_demuxer_get_last_error(demuxerPtr));
            console.error(`Error: Failed to get required input: ${errorMsg}`);
            mp4_file_demuxer_free(demuxerPtr);
            mp4_file_muxer_free(muxerPtr);
            fs.closeSync(inputFile);
            fs.closeSync(outputFile);
            process.exit(1);
        }

        const requiredPosView = new BigUint64Array(memory.buffer, requiredPosPtr, 1);
        const requiredSizeView = new Int32Array(memory.buffer, requiredSizePtr, 1);

        const requiredPos = requiredPosView[0];
        let requiredSize = requiredSizeView[0];

        mp4_free(requiredPosPtr, 8);
        mp4_free(requiredSizePtr, 4);

        if (requiredSize === 0) {
            break;
        }

        let readSize = requiredSize;
        if (requiredSize === -1) {
            readSize = inputSize - Number(requiredPos);
        }

        const buffer = Buffer.alloc(readSize);
        fs.readSync(inputFile, buffer, 0, readSize, Number(requiredPos));

        const wasmBuffer = mp4_alloc(readSize);
        const wasmView = new Uint8Array(memory.buffer, wasmBuffer, readSize);
        wasmView.set(buffer);

        mp4_file_demuxer_handle_input(
            demuxerPtr,
            requiredPos,
            wasmBuffer,
            readSize
        );

        mp4_free(wasmBuffer, readSize);
    }

    // トラック情報を取得
    const tracksPtr = mp4_alloc(8);
    const trackCountPtr = mp4_alloc(4);

    let tracksErr = mp4_file_demuxer_get_tracks(
        demuxerPtr,
        tracksPtr,
        trackCountPtr
    );

    if (tracksErr !== 0) {
        const errorMsg = readCString(mp4_file_demuxer_get_last_error(demuxerPtr));
        console.error(`Error: Failed to get tracks: ${errorMsg}`);
        mp4_file_demuxer_free(demuxerPtr);
        mp4_file_muxer_free(muxerPtr);
        fs.closeSync(inputFile);
        fs.closeSync(outputFile);
        process.exit(1);
    }

    const trackCountView = new Uint32Array(memory.buffer, trackCountPtr, 1);
    const trackCount = trackCountView[0];

    console.log(`Found ${trackCount} track(s)\n`);

    mp4_free(tracksPtr, 8);
    mp4_free(trackCountPtr, 4);

    // ==================== マルチプレックス初期化 ====================

    let err = mp4_file_muxer_initialize(muxerPtr);
    if (err !== 0) {
        const errorMsg = readCString(mp4_file_muxer_get_last_error_func(muxerPtr));
        console.error(`Error: Failed to initialize muxer: ${errorMsg}`);
        mp4_file_demuxer_free(demuxerPtr);
        mp4_file_muxer_free(muxerPtr);
        fs.closeSync(inputFile);
        fs.closeSync(outputFile);
        process.exit(1);
    }

    // マルチプレックサーの初期出力データを書き込む
    let currentOutputOffset = 0;
    const outputOffsetPtr = mp4_alloc(8);
    const outputSizePtr = mp4_alloc(4);
    const outputDataPtrPtr = mp4_alloc(8);

    console.log('Writing initial muxer boxes...');

    while (true) {
        err = mp4_file_muxer_next_output(
            muxerPtr,
            outputOffsetPtr,
            outputSizePtr,
            outputDataPtrPtr
        );

        if (err !== 0) break;

        const offsetView = new BigUint64Array(memory.buffer, outputOffsetPtr, 1);
        const sizeView = new Uint32Array(memory.buffer, outputSizePtr, 1);
        const dataPtrView = new BigUint64Array(memory.buffer, outputDataPtrPtr, 1);

        const offset = offsetView[0];
        const size = sizeView[0];
        const dataPtr = Number(dataPtrView[0]);

        if (size === 0) break;

        const outputData = readU8Array(dataPtr, size);
        fs.writeSync(outputFile, outputData, 0, size, Number(offset));

        console.log(`  Wrote ${size} bytes at offset ${offset}`);
        currentOutputOffset = Number(offset) + size;
    }

    console.log(`Sample data will start at offset: ${currentOutputOffset}\n`);

    mp4_free(outputOffsetPtr, 8);
    mp4_free(outputSizePtr, 4);
    mp4_free(outputDataPtrPtr, 8);

    // ==================== サンプルのリマルチプレックス ====================

    console.log('Remuxing samples...');

    const samplePtr = mp4_alloc(256);
    let sampleCount = 0;

    while (true) {
        const demuxErr = mp4_file_demuxer_next_sample(demuxerPtr, samplePtr);

        if (demuxErr === 8) { // MP4_ERROR_NO_MORE_SAMPLES
            break;
        }

        if (demuxErr !== 0) {
            const errorMsg = readCString(mp4_file_demuxer_get_last_error(demuxerPtr));
            console.error(`Error: Failed to get next sample: ${errorMsg}`);
            break;
        }

        // サンプルデータを読み込む
        const sampleView = new DataView(memory.buffer, samplePtr, 256);
        const dataOffset = Number(sampleView.getBigUint64(24, true));
        const dataSize = sampleView.getUint32(32, true);

        const sampleData = Buffer.alloc(dataSize);
        fs.readSync(inputFile, sampleData, 0, dataSize, Number(dataOffset));

        // 出力ファイルにサンプルデータを追記
        fs.writeSync(outputFile, sampleData, 0, dataSize, currentOutputOffset);

        // JSON からマルチプレックスサンプルを構築
        // （デマルチプレックスサンプルをマルチプレックスサンプルに変換）
        const muxSampleObj = {
            track_kind: 'video', // TODO: 実際にはトラック情報から取得
            sample_entry: null,
            keyframe: true,
            timescale: 30,
            duration: 1,
            data_offset: currentOutputOffset,
            data_size: dataSize,
        };

        const jsonStr = JSON.stringify(muxSampleObj);
        const jsonBytes = new TextEncoder().encode(jsonStr);
        const jsonPtr = writeToMemory(jsonBytes);

        const muxSamplePtr = mp4_mux_sample_from_json(jsonPtr, jsonBytes.length);
        mp4_free(jsonPtr, jsonBytes.length);

        if (muxSamplePtr === 0) {
            console.error('Error: Failed to create mux sample from JSON');
            break;
        }

        // マルチプレックサーにサンプルを追加
        const appendErr = mp4_file_muxer_append_sample(muxerPtr, muxSamplePtr);
        mp4_mux_sample_free(muxSamplePtr);

        if (appendErr !== 0) {
            const errorMsg = readCString(mp4_file_muxer_get_last_error_func(muxerPtr));
            console.error(`Error: Failed to append sample: ${errorMsg}`);
            break;
        }

        sampleCount++;
        currentOutputOffset += dataSize;

        if (sampleCount % 100 === 0) {
            console.log(`  Processed ${sampleCount} samples`);
        }
    }

    console.log(`Total samples processed: ${sampleCount}\n`);

    mp4_free(samplePtr, 256);

    // ==================== マルチプレックス終了処理 ====================

    console.log('Finalizing muxer...');

    err = mp4_file_muxer_finalize(muxerPtr);
    if (err !== 0) {
        const errorMsg = readCString(mp4_file_muxer_get_last_error_func(muxerPtr));
        console.error(`Error: Failed to finalize muxer: ${errorMsg}`);
        mp4_file_demuxer_free(demuxerPtr);
        mp4_file_muxer_free(muxerPtr);
        fs.closeSync(inputFile);
        fs.closeSync(outputFile);
        process.exit(1);
    }

    // マルチプレックサーの最終出力データを書き込む
    const finalOutputOffsetPtr = mp4_alloc(8);
    const finalOutputSizePtr = mp4_alloc(4);
    const finalOutputDataPtrPtr = mp4_alloc(8);

    while (true) {
        err = mp4_file_muxer_next_output(
            muxerPtr,
            finalOutputOffsetPtr,
            finalOutputSizePtr,
            finalOutputDataPtrPtr
        );

        if (err !== 0) break;

        const offsetView = new BigUint64Array(memory.buffer, finalOutputOffsetPtr, 1);
        const sizeView = new Uint32Array(memory.buffer, finalOutputSizePtr, 1);
        const dataPtrView = new BigUint64Array(memory.buffer, finalOutputDataPtrPtr, 1);

        const offset = offsetView[0];
        const size = sizeView[0];
        const dataPtr = Number(dataPtrView[0]);

        if (size === 0) break;

        const outputData = readU8Array(dataPtr, size);
        fs.writeSync(outputFile, outputData, 0, size, Number(offset));

        console.log(`  Wrote final ${size} bytes at offset ${offset}`);
    }

    mp4_free(finalOutputOffsetPtr, 8);
    mp4_free(finalOutputSizePtr, 4);
    mp4_free(finalOutputDataPtrPtr, 8);

    console.log(`\nSuccessfully remuxed '${path.basename(inputPath)}' to '${path.basename(outputPath)}'`);

    // リソースを解放
    mp4_file_muxer_free(muxerPtr);
    mp4_file_demuxer_free(demuxerPtr);
    fs.closeSync(inputFile);
    fs.closeSync(outputFile);
}

// メイン処理
const args = process.argv.slice(2);
if (args.length < 2) {
    console.error('Usage: node remux.js <input.mp4> <output.mp4>');
    process.exit(1);
}

const inputPath = args[0];
const outputPath = args[1];
await remuxMP4File(inputPath, outputPath);

